File: packages/python/src/castia/integrations/toolbox.py
"""Consuming a Microsoft Foundry **Toolbox** from a hosted agent.

A *toolbox* is a curated set of tools the platform exposes behind a single
MCP-compatible endpoint, with centralized auth, governance, and versioning. A
hosted (custom-code) agent is **not** configured with tools in the portal --
your code owns the tool loop -- so the way a toolbox reaches this agent is:

1. Someone creates the toolbox (REST/SDK/CLI) and it gets a stable MCP URL::

       {project_endpoint}/toolboxes/{name}/mcp?api-version=v1

2. ``azd`` injects that address into the container as an environment variable
   (the toolbox *name* embedded in the URL is the only "which toolbox"
   selector -- there is no separate picker).
3. This module reads it at startup and turns it into a single Responses-API
   ``mcp`` tool spec, which the model service resolves **server-side** within a
   turn. Unlike :mod:`castia.tools` function tools, a toolbox tool needs no
   local impl -- the toolbox runs the tool and the model folds the result in.

Server-side ``mcp`` specs do not declare each federated tool as a local
function, but tool wording still matters: Foundry's guidance says descriptions
steer tool choice, and the OpenAI Responses MCP schema exposes
``server_description`` as model-visible context for a remote MCP server. Pass a
``descriptions`` map (and optional ``param_guidance``) when you need local
post-facto wording for selected toolbox tools. The map is folded into
``server_description`` for runtime steering and into a private optimizer sidecar
that :mod:`castia.optimize` serializes to ``tools.json``. This keeps the
validated server-side toolbox call path while making selected federated tools
eligible for Foundry Agent Optimizer description rewrites. Because the builder
is offline and cannot list a remote toolbox, overrides require ``allowed_tools``
so stale names fail loud instead of silently dropping optimizer guidance.
Responses returns MCP calls with ``server_label`` separate from ``name``; live
validation against a Foundry toolbox showed the model-visible name is the
``allowed_tools`` entry (for example ``web``), not ``{server_label}___web``.
Toolbox-federated connection names may still contain their own ``___`` prefix
(for example ``contracts-kb-mcp___knowledge_base_retrieve``).

Env forms are supported in this precedence (see :func:`resolve_toolbox_endpoint`):

* an explicit full URL in ``TOOLBOX_ENDPOINT`` / ``TOOLBOX_MCP_ENDPOINT`` -- a
  passthrough name **we** choose for the container; or
* the **platform-native** ``TOOLBOX_<NORMALIZED_NAME>_MCP_ENDPOINT`` that the
  ``azd ai toolbox`` extension itself writes to the azd environment on
  ``create``/``show`` (normalized = upper-cased, non-alphanumerics -> ``_``),
  resolved from ``TOOLBOX_NAME``; or
* ``FOUNDRY_PROJECT_ENDPOINT`` + ``TOOLBOX_NAME`` (+ optional ``TOOLBOX_VERSION``),
  which this module composes into the same URL. The unversioned URL resolves the
  toolbox's promoted default version, so version bumps need no redeploy.

The builders here are **pure** (offline-testable); :func:`toolbox_token` is the
one impure seam -- it mints an Entra token for the toolbox audience from the
container's managed identity, mirroring :func:`castia.credentials.bot_connector_token`.

.. note::
   Validated live (hal ``hello-world-autopilot`` v40, 2026-09-09): the Responses
   model service **accepts** a Foundry-toolbox ``mcp`` tool authenticated with a
   raw ``https://ai.azure.com`` bearer header minted from the container's managed
   identity -- **no ``project_connection_id`` required**. The turn's span chain
   showed ``tools/list`` + ``tools/call`` against the toolbox succeeding. The
   ``project_connection_id`` path remains supported for stored-connection auth.
"""

from __future__ import annotations

import os
import re
from collections.abc import Mapping
from typing import Any

#: The Entra audience the toolbox MCP endpoint authenticates against. Mirrors the
#: ``azd ai connection create remote-tool ... --audience https://ai.azure.com``
#: guidance for a toolbox remote-tool connection.
AI_FOUNDRY_SCOPE = "https://ai.azure.com/.default"

#: Env var carrying a complete toolbox MCP URL (name embedded). ``TOOLBOX_MCP_ENDPOINT``
#: is accepted as an alias some scaffolds inject.
_FULL_URL_ENVS = ("TOOLBOX_ENDPOINT", "TOOLBOX_MCP_ENDPOINT")
#: Env vars used to *compose* the URL when a full one isn't supplied.
_PROJECT_ENDPOINT_ENV = "FOUNDRY_PROJECT_ENDPOINT"
_NAME_ENV = "TOOLBOX_NAME"
_VERSION_ENV = "TOOLBOX_VERSION"

_API_VERSION = "v1"
_FEDERATED_NAME_SEPARATOR = "___"

# Private castia metadata carried on a raw MCP spec. ``Model.respond_with_tools``
# strips it before calling Responses; ``python -m castia optimize`` consumes it
# when generating ``.agent_configs/baseline/tools.json``.
OPTIMIZER_TOOL_DEFINITIONS_KEY = "x-castia-optimizer-tool-definitions"
_SERVER_DESCRIPTION_KEY = "x-castia-server-description"


def platform_endpoint_env(name: str) -> str:
    """The azd-toolbox-extension env var name for a toolbox's computed MCP URL.

    ``azd ai toolbox create``/``show`` write the endpoint to
    ``TOOLBOX_<NORMALIZED_NAME>_MCP_ENDPOINT`` in the active azd environment (and
    ``delete`` clears it). Normalization upper-cases the name and replaces every
    run of non-alphanumeric characters with a single underscore, so ``hal-smoke``
    -> ``TOOLBOX_HAL_SMOKE_MCP_ENDPOINT``.
    """
    slug = re.sub(r"[^0-9A-Za-z]+", "_", name).strip("_").upper()
    return f"TOOLBOX_{slug}_MCP_ENDPOINT"


def compose_toolbox_endpoint(
    project_endpoint: str, name: str, *, version: str | None = None
) -> str:
    """Build the toolbox MCP URL from a project endpoint and a toolbox name.

    ``version`` selects a pinned ``/versions/{version}/`` URL; omit it for the
    unversioned URL, which resolves the promoted default version.
    """
    base = project_endpoint.rstrip("/")
    if version:
        path = f"/toolboxes/{name}/versions/{version}/mcp"
    else:
        path = f"/toolboxes/{name}/mcp"
    return f"{base}{path}?api-version={_API_VERSION}"


def resolve_toolbox_endpoint(env: Mapping[str, str] | None = None) -> str | None:
    """Resolve the toolbox MCP URL from the environment, or ``None`` if unset.

    Precedence:

    1. an explicit full URL in ``TOOLBOX_ENDPOINT`` / ``TOOLBOX_MCP_ENDPOINT``
       (a passthrough var we plumb into the container ourselves);
    2. the **platform-native** ``TOOLBOX_<NORMALIZED_NAME>_MCP_ENDPOINT`` the azd
       toolbox extension writes, keyed off ``TOOLBOX_NAME`` (see
       :func:`platform_endpoint_env`) -- so you can forward the extension's own
       var verbatim;
    3. compose from ``FOUNDRY_PROJECT_ENDPOINT`` + ``TOOLBOX_NAME`` (+ optional
       ``TOOLBOX_VERSION``).

    Returns ``None`` when none is available so a caller can treat "no toolbox
    configured" as simply attaching no tool.
    """
    env = os.environ if env is None else env
    for key in _FULL_URL_ENVS:
        url = env.get(key)
        if url:
            return url
    name = env.get(_NAME_ENV)
    if name:
        platform_url = env.get(platform_endpoint_env(name))
        if platform_url:
            return platform_url
        project = env.get(_PROJECT_ENDPOINT_ENV)
        if project:
            return compose_toolbox_endpoint(
                project, name, version=env.get(_VERSION_ENV) or None
            )
    return None


def toolbox_mcp_tool(
    endpoint: str | None = None,
    *,
    env: Mapping[str, str] | None = None,
    server_label: str = "toolbox",
    allowed_tools: list[str] | tuple[str, ...] | None = None,
    require_approval: str = "never",
    token: str | None = None,
    project_connection_id: str | None = None,
    headers: Mapping[str, str] | None = None,
    server_description: str | None = None,
    descriptions: Mapping[str, str] | None = None,
    param_guidance: Mapping[str, Mapping[str, str]] | None = None,
) -> dict[str, Any] | None:
    """A Responses-API ``mcp`` tool spec for a Foundry toolbox, or ``None``.

    ``endpoint`` defaults to :func:`resolve_toolbox_endpoint` over ``env`` (the
    process environment by default), so in a deployed agent this is called with
    no arguments and reads ``TOOLBOX_*`` -- returning ``None`` (attach nothing)
    when no toolbox is configured.

    Auth is either a bearer ``token`` (folded into an ``Authorization`` header for
    the model service to present to the toolbox) or a ``project_connection_id``
    (the toolbox resolves auth from a stored project connection). Extra ``headers``
    are merged last. ``allowed_tools`` restricts which of the toolbox's tools the
    model may call; omit it to expose all.

    ``descriptions`` and ``param_guidance`` are local, post-facto guidance for
    selected federated tools. Keys may be the final model-visible name
    (``{server_label}___{tool}``), the selected upstream name, or the bare tool
    name after the final ``___`` segment. The legacy ``{server_label}___{tool}``
    spelling is accepted as an alias but normalizes back to the selected tool
    name. Unknown or ambiguous keys raise. Since the builder cannot discover a
    remote toolbox offline, ``allowed_tools`` is required whenever overrides are
    supplied.
    """
    override_defs = _optimizer_tool_definitions(
        server_label=server_label,
        allowed_tools=allowed_tools,
        descriptions=descriptions,
        param_guidance=param_guidance,
    )
    endpoint = endpoint or resolve_toolbox_endpoint(env)
    if not endpoint:
        return None
    spec: dict[str, Any] = {
        "type": "mcp",
        "server_label": server_label,
        "server_url": endpoint,
        "require_approval": require_approval,
    }
    if allowed_tools:
        spec["allowed_tools"] = list(allowed_tools)
    if server_description or override_defs:
        if server_description:
            spec[_SERVER_DESCRIPTION_KEY] = server_description
        spec["server_description"] = _server_description(
            server_description, override_defs
        )
    if override_defs:
        spec[OPTIMIZER_TOOL_DEFINITIONS_KEY] = override_defs
    if project_connection_id:
        spec["project_connection_id"] = project_connection_id
    merged: dict[str, str] = {}
    if token:
        # Concatenate rather than interpolate: the workspace secret-redaction
        # filter rewrites an interpolated-bearer literal on save. See
        # castia.credentials.bearer for the same guard.
        merged["Authorization"] = "Bearer " + token
    if headers:
        merged.update(headers)
    if merged:
        spec["headers"] = merged
    return spec


def knowledge_base_mcp_tool(
    endpoint: str,
    *,
    search_token: str | None = None,
    allowed_tools: list[str] | tuple[str, ...] = ("knowledge_base_retrieve",),
    server_label: str = "knowledge-base",
    require_approval: str = "never",
    token: str | None = None,
    project_connection_id: str | None = None,
    server_description: str | None = None,
    descriptions: Mapping[str, str] | None = None,
    param_guidance: Mapping[str, Mapping[str, str]] | None = None,
) -> dict[str, Any]:
    """A Responses-API ``mcp`` tool spec for a **Foundry IQ** knowledge base.

    Foundry IQ knowledge bases are their own MCP server (Azure AI Search agentic
    retrieval): ``{search}/knowledgebases/{kb}/mcp?api-version=...``. They accept
    a per-user Azure AI Search token forwarded as ``x-ms-query-source-authorization``
    -- the same header a managed prompt agent templates from ``structured_inputs``.
    Pass it as ``search_token``. ``token`` (an ``Authorization`` bearer) and
    ``project_connection_id`` carry the outer connection auth, if used. Override
    descriptions follow :func:`toolbox_mcp_tool` and default to resolving against
    ``knowledge_base_retrieve``.
    """
    headers: dict[str, str] = {}
    if search_token:
        headers["x-ms-query-source-authorization"] = search_token
    spec = toolbox_mcp_tool(
        endpoint,
        server_label=server_label,
        allowed_tools=allowed_tools,
        require_approval=require_approval,
        token=token,
        project_connection_id=project_connection_id,
        headers=headers or None,
        server_description=server_description,
        descriptions=descriptions,
        param_guidance=param_guidance,
    )
    # endpoint is a required non-empty argument here, so toolbox_mcp_tool never
    # returns None; assert to satisfy the type checker and fail loud on misuse.
    assert spec is not None
    return spec


def _optimizer_tool_definitions(
    *,
    server_label: str,
    allowed_tools: list[str] | tuple[str, ...] | None,
    descriptions: Mapping[str, str] | None,
    param_guidance: Mapping[str, Mapping[str, str]] | None,
) -> list[dict[str, Any]]:
    descriptions = descriptions or {}
    param_guidance = param_guidance or {}
    if not descriptions and not param_guidance:
        return []
    if not allowed_tools:
        raise ValueError(
            "allowed_tools is required when toolbox descriptions or "
            "param_guidance are provided; castia cannot validate override names "
            "without a selected toolbox tool list"
        )

    selected = [_SelectedTool(server_label, name) for name in allowed_tools]
    resolved_descriptions = _resolve_overrides("descriptions", descriptions, selected)
    resolved_params = _resolve_overrides("param_guidance", param_guidance, selected)

    out: list[dict[str, Any]] = []
    for tool in selected:
        description = resolved_descriptions.get(tool.final_name)
        params = resolved_params.get(tool.final_name, {})
        if description is None and not params:
            continue
        out.append(
            {
                "type": "function",
                "function": {
                    "name": tool.final_name,
                    "description": description
                    or f"Call the federated toolbox tool {tool.upstream_name!r}.",
                    "parameters": _parameter_schema(params),
                },
            }
        )
    return out


class _SelectedTool:
    def __init__(self, server_label: str, upstream_name: str) -> None:
        self.upstream_name = upstream_name
        self.final_name = upstream_name
        self.legacy_labeled_name = (
            f"{server_label}{_FEDERATED_NAME_SEPARATOR}{upstream_name}"
        )
        self.bare_name = upstream_name.rsplit(_FEDERATED_NAME_SEPARATOR, 1)[-1]

    def aliases(self) -> set[str]:
        return {
            self.final_name,
            self.upstream_name,
            self.legacy_labeled_name,
            self.bare_name,
        }


def _resolve_overrides(
    label: str, overrides: Mapping[str, Any], selected: list[_SelectedTool]
) -> dict[str, Any]:
    resolved: dict[str, Any] = {}
    for key, value in overrides.items():
        matches = [tool for tool in selected if key in tool.aliases()]
        if not matches:
            allowed = ", ".join(tool.final_name for tool in selected)
            raise ValueError(
                f"unknown toolbox {label} key {key!r}; expected one of: {allowed}"
            )
        if len(matches) > 1:
            choices = ", ".join(tool.final_name for tool in matches)
            raise ValueError(
                f"ambiguous toolbox {label} key {key!r}; matches: {choices}"
            )
        resolved[matches[0].final_name] = value
    return resolved


def _parameter_schema(guidance: Mapping[str, str]) -> dict[str, Any]:
    return {
        "type": "object",
        "properties": {
            name: {"description": description}
            for name, description in guidance.items()
        },
        "required": [],
        "additionalProperties": True,
    }


def _server_description(
    base: str | None, tool_definitions: list[dict[str, Any]]
) -> str:
    lines: list[str] = []
    if base:
        lines.append(base)
    if tool_definitions:
        lines.append("Local tool guidance:")
        for item in tool_definitions:
            func = item["function"]
            lines.append(f"- {func['name']}: {func['description']}")
            props = func.get("parameters", {}).get("properties", {})
            for name, schema in props.items():
                if schema.get("description"):
                    lines.append(f"  - {name}: {schema['description']}")
    return "\n".join(lines)


def apply_optimized_toolbox_tools(
    spec: dict[str, Any], tool_definitions: tuple[dict, ...] | list[dict]
) -> dict[str, Any]:
    """Apply optimizer-rewritten toolbox descriptions to a raw MCP spec.

    Matches candidate ``tools.json`` entries by the optimizer sidecar's function
    names, rewrites only descriptions and parameter-description text, and
    regenerates ``server_description`` so candidate runs can actually exercise
    the rewritten wording while preserving the validated server-side MCP runtime.
    Specs without a toolbox sidecar pass through unchanged.
    """
    current = spec.get(OPTIMIZER_TOOL_DEFINITIONS_KEY)
    if not isinstance(current, list) or not current or not tool_definitions:
        return spec

    lookup: dict[str, dict] = {}
    for item in tool_definitions:
        if not isinstance(item, dict):
            continue
        func = item.get("function")
        if not isinstance(func, dict):
            func = item if item.get("name") else {}
        name = func.get("name")
        if name:
            lookup[name] = func

    if not lookup:
        return spec

    changed = False
    rewritten: list[dict[str, Any]] = []
    for item in current:
        if not isinstance(item, dict):
            rewritten.append(item)
            continue
        func = item.get("function")
        if not isinstance(func, dict):
            rewritten.append(item)
            continue
        optimized = lookup.get(func.get("name"))
        if optimized is None:
            rewritten.append(item)
            continue

        new_func = dict(func)
        description = optimized.get("description")
        if description and description != func.get("description"):
            new_func["description"] = description
            changed = True

        parameters = _merge_parameter_descriptions(
            new_func.get("parameters"), optimized.get("parameters")
        )
        if parameters is not new_func.get("parameters"):
            new_func["parameters"] = parameters
            changed = True

        rewritten.append({**item, "function": new_func})

    if not changed:
        return spec

    base = spec.get(_SERVER_DESCRIPTION_KEY)
    if not isinstance(base, str):
        base = None
    out = {**spec, OPTIMIZER_TOOL_DEFINITIONS_KEY: rewritten}
    out["server_description"] = _server_description(base, rewritten)
    return out


def _merge_parameter_descriptions(current: object, optimized: object | None) -> object:
    if not isinstance(current, dict) or not isinstance(optimized, dict):
        return current
    cur_props = current.get("properties")
    opt_props = optimized.get("properties")
    if not isinstance(cur_props, dict) or not isinstance(opt_props, dict):
        return current

    merged_props: dict[str, Any] = {}
    changed = False
    for name, schema in cur_props.items():
        opt_schema = opt_props.get(name)
        if (
            isinstance(schema, dict)
            and isinstance(opt_schema, dict)
            and opt_schema.get("description")
            and opt_schema.get("description") != schema.get("description")
        ):
            merged_props[name] = {**schema, "description": opt_schema["description"]}
            changed = True
        else:
            merged_props[name] = schema

    if not changed:
        return current
    return {**current, "properties": merged_props}


async def toolbox_token(scope: str = AI_FOUNDRY_SCOPE) -> str:
    """Mint an Entra token for the toolbox audience from the container identity.

    The hosted agent authenticates to its toolbox as its own Entra identity;
    :class:`~azure.identity.aio.DefaultAzureCredential` resolves the container's
    managed identity in Foundry (and a developer login locally). Mirrors
    :func:`castia.credentials.bot_connector_token`. Impure (network) -- kept
    out of the unit-tested builders above.
    """
    from azure.identity.aio import DefaultAzureCredential

    credential = DefaultAzureCredential()
    try:
        token = await credential.get_token(scope)
    finally:
        await credential.close()
    return token.token

File: packages/rust/src/integrations/toolbox.rs
use crate::model::IntegrationsToolboxRuntime;
use crate::optimizing::OPTIMIZER_TOOL_DEFINITIONS_KEY;
use regex::Regex;
use serde_json::{json, Map, Value};
use std::sync::OnceLock;

pub const AI_FOUNDRY_SCOPE: &str = "https://ai.azure.com/.default";
pub const SERVER_DESCRIPTION_KEY: &str = "x-castia-server-description";
const API_VERSION: &str = "v1";
const FEDERATED_NAME_SEPARATOR: &str = "___";

#[derive(Debug, Clone)]
pub struct ToolboxError(String);

impl std::fmt::Display for ToolboxError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.write_str(&self.0)
    }
}

impl std::error::Error for ToolboxError {}

#[derive(Debug, Default, Clone, Copy)]
pub struct CastiaIntegrationsToolboxRuntime;

#[async_trait::async_trait]
impl IntegrationsToolboxRuntime for CastiaIntegrationsToolboxRuntime {
    fn compose_toolbox_endpoint(
        &self,
        project_endpoint: &String,
        name: &String,
        version: &Option<String>,
    ) -> String {
        compose_toolbox_endpoint(project_endpoint, name, version.as_deref())
    }

    fn knowledge_base_mcp_tool(&self, config: &Value) -> Value {
        knowledge_base_mcp_tool(config).unwrap_or_else(|error| panic!("{}", error.0))
    }

    fn platform_endpoint_env(&self, name: &String) -> String {
        platform_endpoint_env(name)
    }

    fn resolve_toolbox_endpoint(&self, env: &Value) -> Option<String> {
        resolve_toolbox_endpoint(env)
    }

    fn toolbox_mcp_tool(&self, config: &Value) -> Value {
        toolbox_mcp_tool(config).unwrap_or_else(|error| panic!("{}", error.0))
    }
}

pub fn platform_endpoint_env(name: &str) -> String {
    static NON_ALNUM: OnceLock<Regex> = OnceLock::new();
    let slug = NON_ALNUM
        .get_or_init(|| Regex::new(r"[^0-9A-Za-z]+").expect("valid regex"))
        .replace_all(name, "_")
        .trim_matches('_')
        .to_ascii_uppercase();
    format!("TOOLBOX_{slug}_MCP_ENDPOINT")
}

pub fn compose_toolbox_endpoint(
    project_endpoint: &str,
    name: &str,
    version: Option<&str>,
) -> String {
    let base = project_endpoint.trim_end_matches('/');
    let path = if let Some(version) = version.filter(|value| !value.is_empty()) {
        format!("/toolboxes/{name}/versions/{version}/mcp")
    } else {
        format!("/toolboxes/{name}/mcp")
    };
    format!("{base}{path}?api-version={API_VERSION}")
}

pub fn resolve_toolbox_endpoint(env: &Value) -> Option<String> {
    let env = env.as_object()?;
    for key in ["TOOLBOX_ENDPOINT", "TOOLBOX_MCP_ENDPOINT"] {
        if let Some(url) = env
            .get(key)
            .and_then(Value::as_str)
            .filter(|value| !value.is_empty())
        {
            return Some(url.to_string());
        }
    }
    let name = env
        .get("TOOLBOX_NAME")
        .and_then(Value::as_str)
        .filter(|value| !value.is_empty())?;
    let platform = platform_endpoint_env(name);
    if let Some(url) = env
        .get(&platform)
        .and_then(Value::as_str)
        .filter(|value| !value.is_empty())
    {
        return Some(url.to_string());
    }
    env.get("FOUNDRY_PROJECT_ENDPOINT")
        .and_then(Value::as_str)
        .filter(|value| !value.is_empty())
        .map(|project| {
            compose_toolbox_endpoint(
                project,
                name,
                env.get("TOOLBOX_VERSION").and_then(Value::as_str),
            )
        })
}

pub fn toolbox_mcp_tool(config: &Value) -> Result<Value, ToolboxError> {
    let object = config.as_object().cloned().unwrap_or_default();
    let allowed_tools = string_array(
        object
            .get("allowed_tools")
            .or_else(|| object.get("allowedTools")),
    )?;
    let descriptions = object
        .get("descriptions")
        .and_then(Value::as_object)
        .cloned()
        .unwrap_or_default();
    let param_guidance = object
        .get("param_guidance")
        .or_else(|| object.get("paramGuidance"))
        .and_then(Value::as_object)
        .cloned()
        .unwrap_or_default();
    let server_label = object
        .get("server_label")
        .or_else(|| object.get("serverLabel"))
        .and_then(Value::as_str)
        .unwrap_or("toolbox");
    let override_defs =
        optimizer_tool_definitions(server_label, &allowed_tools, &descriptions, &param_guidance)?;
    let endpoint = object
        .get("endpoint")
        .and_then(Value::as_str)
        .map(str::to_string)
        .or_else(|| resolve_toolbox_endpoint(object.get("env").unwrap_or(&Value::Null)));
    let Some(endpoint) = endpoint.filter(|value| !value.is_empty()) else {
        return Ok(Value::Null);
    };
    let mut spec = Map::new();
    spec.insert("type".to_string(), Value::String("mcp".to_string()));
    spec.insert(
        "server_label".to_string(),
        Value::String(server_label.to_string()),
    );
    spec.insert("server_url".to_string(), Value::String(endpoint));
    spec.insert(
        "require_approval".to_string(),
        Value::String(
            object
                .get("require_approval")
                .or_else(|| object.get("requireApproval"))
                .and_then(Value::as_str)
                .unwrap_or("never")
                .to_string(),
        ),
    );
    if !allowed_tools.is_empty() {
        spec.insert("allowed_tools".to_string(), strings(&allowed_tools));
    }
    let base_description = object
        .get("server_description")
        .or_else(|| object.get("serverDescription"))
        .and_then(Value::as_str)
        .filter(|value| !value.is_empty());
    if base_description.is_some() || !override_defs.is_empty() {
        if let Some(base) = base_description {
            spec.insert(
                SERVER_DESCRIPTION_KEY.to_string(),
                Value::String(base.to_string()),
            );
        }
        spec.insert(
            "server_description".to_string(),
            Value::String(server_description(base_description, &override_defs)),
        );
    }
    if !override_defs.is_empty() {
        spec.insert(
            OPTIMIZER_TOOL_DEFINITIONS_KEY.to_string(),
            Value::Array(override_defs),
        );
    }
    if let Some(connection) = object
        .get("project_connection_id")
        .or_else(|| object.get("projectConnectionId"))
        .and_then(Value::as_str)
        .filter(|value| !value.is_empty())
    {
        spec.insert(
            "project_connection_id".to_string(),
            Value::String(connection.to_string()),
        );
    }
    let mut headers = object
        .get("headers")
        .and_then(Value::as_object)
        .cloned()
        .unwrap_or_default();
    if let Some(token) = object
        .get("token")
        .and_then(Value::as_str)
        .filter(|value| !value.is_empty())
    {
        let mut merged = Map::new();
        merged.insert(
            "Authorization".to_string(),
            Value::String(format!("Bearer {token}")),
        );
        merged.extend(headers);
        headers = merged;
    }
    if !headers.is_empty() {
        spec.insert("headers".to_string(), Value::Object(headers));
    }
    Ok(Value::Object(spec))
}

pub fn knowledge_base_mcp_tool(config: &Value) -> Result<Value, ToolboxError> {
    let object = config.as_object().cloned().unwrap_or_default();
    let endpoint = object
        .get("endpoint")
        .and_then(Value::as_str)
        .filter(|value| !value.is_empty())
        .ok_or_else(|| ToolboxError("endpoint is required".to_string()))?;
    let mut forwarded = object.clone();
    forwarded.insert("endpoint".to_string(), Value::String(endpoint.to_string()));
    forwarded.insert(
        "server_label".to_string(),
        Value::String(
            object
                .get("server_label")
                .or_else(|| object.get("serverLabel"))
                .and_then(Value::as_str)
                .unwrap_or("knowledge-base")
                .to_string(),
        ),
    );
    if !forwarded.contains_key("allowed_tools") && !forwarded.contains_key("allowedTools") {
        forwarded.insert(
            "allowed_tools".to_string(),
            json!(["knowledge_base_retrieve"]),
        );
    }
    let mut headers = object
        .get("headers")
        .and_then(Value::as_object)
        .cloned()
        .unwrap_or_default();
    if let Some(token) = object
        .get("search_token")
        .or_else(|| object.get("searchToken"))
        .and_then(Value::as_str)
        .filter(|value| !value.is_empty())
    {
        headers.insert(
            "x-ms-query-source-authorization".to_string(),
            Value::String(token.to_string()),
        );
    }
    if !headers.is_empty() {
        forwarded.insert("headers".to_string(), Value::Object(headers));
    }
    toolbox_mcp_tool(&Value::Object(forwarded))
}

fn optimizer_tool_definitions(
    server_label: &str,
    allowed_tools: &[String],
    descriptions: &Map<String, Value>,
    param_guidance: &Map<String, Value>,
) -> Result<Vec<Value>, ToolboxError> {
    if descriptions.is_empty() && param_guidance.is_empty() {
        return Ok(Vec::new());
    }
    if allowed_tools.is_empty() {
        return Err(ToolboxError(
            "allowed_tools is required when toolbox descriptions or param_guidance are provided; castia cannot validate override names without a selected toolbox tool list".to_string(),
        ));
    }
    let selected = allowed_tools
        .iter()
        .map(|name| SelectedTool::new(server_label, name))
        .collect::<Vec<_>>();
    let resolved_descriptions = resolve_overrides("descriptions", descriptions, &selected)?;
    let resolved_params = resolve_overrides("param_guidance", param_guidance, &selected)?;
    let mut out = Vec::new();
    for tool in selected {
        let has_description_override = resolved_descriptions.contains_key(&tool.final_name);
        let description = resolved_descriptions
            .get(&tool.final_name)
            .and_then(Value::as_str)
            .filter(|value| !value.is_empty());
        let params = resolved_params
            .get(&tool.final_name)
            .and_then(Value::as_object)
            .cloned()
            .unwrap_or_default();
        if !has_description_override && params.is_empty() {
            continue;
        }
        out.push(json!({
            "type": "function",
            "function": {
                "name": tool.final_name,
                "description": description.map(str::to_string).unwrap_or_else(|| format!("Call the federated toolbox tool '{}'.", tool.upstream_name)),
                "parameters": parameter_schema(&params),
            },
        }));
    }
    Ok(out)
}

#[derive(Clone)]
struct SelectedTool {
    upstream_name: String,
    final_name: String,
    legacy_labeled_name: String,
    bare_name: String,
}

impl SelectedTool {
    fn new(server_label: &str, upstream_name: &str) -> Self {
        Self {
            upstream_name: upstream_name.to_string(),
            final_name: upstream_name.to_string(),
            legacy_labeled_name: format!("{server_label}{FEDERATED_NAME_SEPARATOR}{upstream_name}"),
            bare_name: upstream_name
                .rsplit(FEDERATED_NAME_SEPARATOR)
                .next()
                .unwrap_or(upstream_name)
                .to_string(),
        }
    }

    fn matches(&self, key: &str) -> bool {
        key == self.final_name
            || key == self.upstream_name
            || key == self.legacy_labeled_name
            || key == self.bare_name
    }
}

fn resolve_overrides(
    label: &str,
    overrides: &Map<String, Value>,
    selected: &[SelectedTool],
) -> Result<Map<String, Value>, ToolboxError> {
    let mut resolved = Map::new();
    for (key, value) in overrides {
        let matches = selected
            .iter()
            .filter(|tool| tool.matches(key))
            .collect::<Vec<_>>();
        if matches.is_empty() {
            let allowed = selected
                .iter()
                .map(|tool| tool.final_name.as_str())
                .collect::<Vec<_>>()
                .join(", ");
            return Err(ToolboxError(format!(
                "unknown toolbox {label} key '{key}'; expected one of: {allowed}"
            )));
        }
        if matches.len() > 1 {
            let choices = matches
                .iter()
                .map(|tool| tool.final_name.as_str())
                .collect::<Vec<_>>()
                .join(", ");
            return Err(ToolboxError(format!(
                "ambiguous toolbox {label} key '{key}'; matches: {choices}"
            )));
        }
        resolved.insert(matches[0].final_name.clone(), value.clone());
    }
    Ok(resolved)
}

fn parameter_schema(guidance: &Map<String, Value>) -> Value {
    let properties = guidance
        .iter()
        .filter_map(|(name, description)| {
            description
                .as_str()
                .map(|description| (name.clone(), json!({"description": description})))
        })
        .collect::<Map<_, _>>();
    json!({
        "type": "object",
        "properties": properties,
        "required": [],
        "additionalProperties": true,
    })
}

fn server_description(base: Option<&str>, tool_definitions: &[Value]) -> String {
    let mut lines = Vec::new();
    if let Some(base) = base.filter(|value| !value.is_empty()) {
        lines.push(base.to_string());
    }
    if !tool_definitions.is_empty() {
        lines.push("Local tool guidance:".to_string());
        for item in tool_definitions {
            let Some(func) = item.get("function").and_then(Value::as_object) else {
                continue;
            };
            let name = func.get("name").and_then(Value::as_str).unwrap_or_default();
            let description = func
                .get("description")
                .and_then(Value::as_str)
                .unwrap_or_default();
            lines.push(format!("- {name}: {description}"));
            if let Some(props) = func
                .get("parameters")
                .and_then(|value| value.get("properties"))
                .and_then(Value::as_object)
            {
                for (name, schema) in props {
                    if let Some(description) = schema.get("description").and_then(Value::as_str) {
                        lines.push(format!("  - {name}: {description}"));
                    }
                }
            }
        }
    }
    lines.join("\n")
}

fn string_array(value: Option<&Value>) -> Result<Vec<String>, ToolboxError> {
    match value {
        None | Some(Value::Null) => Ok(Vec::new()),
        Some(Value::Array(items)) => items
            .iter()
            .map(|item| {
                item.as_str().map(str::to_string).ok_or_else(|| {
                    ToolboxError("allowed_tools must be a list of strings".to_string())
                })
            })
            .collect(),
        _ => Err(ToolboxError(
            "allowed_tools must be a list of strings".to_string(),
        )),
    }
}

fn strings(values: &[String]) -> Value {
    Value::Array(values.iter().cloned().map(Value::String).collect())
}
